File: Javascript/1.Agosto/10.3Strikes-20Agosto.js
//todo Dato un numero intero compreso tra 1 e 10.000, restituisci un conteggio di quanti numeri da 1 fino a quell'intero il cui quadrato contiene almeno una cifra 3.

//? Tempo di esecuzione:
//* Non risolto (per poco)

//?Codice scritto senza utilizzo di ChatGpt

function squaresWithThree(n) {
  if(n === 1) return 0;

  let contatore = 0

  for(let i = 0; i < n; i++){
    let squares = i * i;
    for(let j = 0; j < squares.length; j++){
      if(squares[j] === 3)
      contatore++
    }
  }
  return contatore;
}

console.log(squaresWithThree(1000))

//?Codice scritto da ChatGpt

function squaresWithThree(n) {
  let count = 0;

  for (let i = 1; i <= n; i++) { //! Chat fa partire il ciclo direttamente da 1, così da evitrare la condizione iniziale che ho messo io
    const squareStr = (i * i).toString(); //! Ecco dove stava l'errore! squares è un numero, non ha .length
    if (squareStr.includes("3")) { //! Avevo pensato all'includes, ma non riuscivo ad appliccarlo come volessi 
      count++;
    }
  }

  return count;
}

//* Hai capito cosa calcolare (quadrati)

//* Struttura a cicli corretta

//* Tentativo giusto di analizzare le cifre

//! squares.length su un numero (errore concettuale)

//! Doppio ciclo non necessario

//! Logica non allineata alla consegna (“almeno un 3”)

//todo Considerazioni: 

//* L'esercizio mi ha messo difficoltà nel capire, però è stato utile

//! N.B. BASTA FOR IN FOR NON NECESSARI

File: Python/1.Agosto/10.3Strikes-20Agosto.py
# ============================================================
# 🧠 ESERCIZIO
# ============================================================

# Dato un numero intero n (1 → 10.000),
# contare quanti numeri da 1 a n hanno il quadrato
# che contiene almeno una cifra "3".

# Esempio:
# 10 → controlli 1², 2², ... 10²
# e conti quelli che hanno "3" nel risultato


# ============================================================
# 👨‍💻 CODICE SCRITTO DA ME
# ============================================================

def squares_with_three(n):

    if n == 1: return 0

    cont = 0;

    for number in range(1, n + 1):
        quadrato = number ** 2
        if "3" in str(quadrato):
            cont += 1

    return cont


print(squares_with_three(100))


# ============================================================
# ⚡ FEEDBACK VELOCE
# ============================================================

# Codice corretto 👍
# Logica chiara e funzionante
# Ottimo uso di:
# - range
# - conversione str()
# - controllo "in"

# Nessun errore logico importante


# ============================================================
# ❌ ERRORI SPIEGATI
# ============================================================

# ⚠️ if n == 1: return 0
# Non è necessario e può essere sbagliato in altri contesti

# Perché?
# 1² = 1 → non contiene "3" → quindi 0 è corretto
# ma il controllo è ridondante

# ✔️ meglio lasciare la logica generale


# ❌ punto e virgola ";"
# non serve in Python


# ============================================================
# ✅ CODICE CORRETTO (PULITO)
# ============================================================

def squares_with_three(n):

    cont = 0

    for number in range(1, n + 1):

        quadrato = number ** 2

        # trasformo in stringa e cerco la cifra "3"
        if "3" in str(quadrato):
            cont += 1

    return cont


# ============================================================
# 📖 SPIEGAZIONE RIGA PER RIGA
# ============================================================

# cont = 0
# → contatore dei numeri validi

# for number in range(1, n + 1):
# → ciclo su tutti i numeri da 1 a n

# quadrato = number ** 2
# → calcolo il quadrato

# str(quadrato)
# → converto il numero in stringa
# esempio: 49 → "49"

# "3" in str(quadrato)
# → controllo se contiene la cifra 3

# cont += 1
# → incremento il contatore


# ============================================================
# 🧪 TEST
# ============================================================

print(squares_with_three(10))
print(squares_with_three(100))
print(squares_with_three(500))